ext4: Fix hang on umount with quotas when journal is aborted

From: Jan Kara <jack@suse.cz>

Call dquot_drop() from ext4_dquot_drop() even if we fail to start a
transaction. Otherwise we never get to dropping references to quota structures
from the inode and umount will hang indefinitely.  Thanks to Payphone LIOU for
spotting the problem.

Signed-off-by: Jan Kara <jack@suse.cz>
Signed-off-by: Mingming Cao <cmm@us.ibm.com>
CC: Payphone LIOU <lioupayphone@gmail.com>
Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>

---
  I'm not sure if this should go through ext4 patch queue or through Andrew.
It should not collide with anything so it probably does not matter much...

 fs/ext4/super.c |    8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

Index: linux-2.6.25-rc9/fs/ext4/super.c
===================================================================
--- linux-2.6.25-rc9.orig/fs/ext4/super.c	2008-04-14 11:12:45.000000000 -0700
+++ linux-2.6.25-rc9/fs/ext4/super.c	2008-04-14 11:12:48.000000000 -0700
@@ -3038,8 +3038,14 @@ static int ext4_dquot_drop(struct inode 
 
 	/* We may delete quota structure so we need to reserve enough blocks */
 	handle = ext4_journal_start(inode, 2*EXT4_QUOTA_DEL_BLOCKS(inode->i_sb));
-	if (IS_ERR(handle))
+	if (IS_ERR(handle)) {
+		/*
+		 * We call dquot_drop() anyway to at least release references
+		 * to quota structures so that umount does not hang.
+		 */
+		dquot_drop(inode);
 		return PTR_ERR(handle);
+	}
 	ret = dquot_drop(inode);
 	err = ext4_journal_stop(handle);
 	if (!ret)
